python/lsst/daf/persistence/policy.py

```diff
@@ -0,0 +1,252 @@
+    That is, foo.getValue('a.b.c') is the same as foo['a']['b']['c'] is the same as foo['a.b.c'], and either
+    of these syntaxes may be used.
+
+    Storage formats supported:
+    - yaml: read and write is supported.
+    - pex policy: read is supported, although this is deprecated and will at some point be removed.
+    """
+
+    def __getitem__ (self, name):
+        data = self.data
+        for key in name.split('.'):
+            if key in data:
+                data = data[key]
+            else:
+                return None
+        if isinstance(data, collections.Mapping):
+            data = Policy(data)
+        return data
+
+    def __setitem__(self, name, value):
+        if isinstance(value, collections.Mapping):
+            keys = name.split('.')
+            d = {}
+            cur = d
+            for key in keys[0:-1]:
+                cur[key] = {}
+                cur = cur[key]
+            cur[keys[-1]] = value
+            self.update(d)
+        data = self.data
+        keys = name.split('.')
+        for key in keys[0:-1]:
+            data = data.setdefault(key, {})
+        data[keys[-1]] = value
+
+    def __contains__(self, key):
+        d = self.data
+        keys = key.split('.')
+        for k in keys[0:-1]:
+            if k in d:
+                d = d[k]
+            else:
+                return False
+        return keys[-1] in d
+
+    def update(self, other):
+        """Like dict.update, but will add or modify keys in nested dicts, instead of overwriting the nested
+        dict entirely.
+
+        For example, for the given code:
+        foo = {'a': {'b': 1}}
+        foo.update({'a': {'c': 2}})
+
+        If foo is a dict, then after the update foo == {'a': {'c': 2}}
+        But if foo is a Policy, then after the update foo == {'a': {'b': 1, 'c': 2}}
+        """
+        def doUpdate(d, u):
+            for k, v in u.iteritems():
+                if isinstance(d, collections.Mapping):
+                    if isinstance(v, collections.Mapping):
+                        r = doUpdate(d.get(k, {}), v)
+                        d[k] = r
+                    else:
+                        d[k] = u[k]
+                else:
+                    d = {k: u[k]}
+            return d
+        doUpdate(self.data, other)
+
+
+    #########################
+    # mimics pex_policy api #
+    #########################
+
+    def names(self, topLevelOnly=False):
+        """Get the dot-delimited name of all the keys in the hierarchy.
+        """
+        if topLevelOnly:
+            return self.keys()
+
+        def getKeys(d, keys, base):
+            for key in d.keys():
+                val = d[key]
+                levelKey = base + '.' + key if base is not None else key
+                keys.append(levelKey)
+                if isinstance(val, collections.Mapping):
+                    getKeys(val, keys, levelKey)
+        keys = []
+        getKeys(self.data, keys, None)
+        return keys
+
```

Comment:
I'd like everything from here down to disappear eventually, if not now.  The idea is that we should use native Python interfaces rather than our own C++-like wrappers.